### PR TITLE
component_nonce variable not updated properly on node restart.

### LINKED ISSUE #5133: Component nonce is not updated properly on node restart.
In our current implementation we store all edges removed to DB. Each set of edges is called `component` and has unique id `nonce`.

There is currently a bug, where `nonce` that's saved in database doesn't get updated properly.

If nonce is currently let's say `self.component_nonce = 10`.  In function `try_save_edges`,
which creates new component, we will increase this value by 1; save to database `component_nonce = 10` ;
assign new component to have id `10`.

If restart happens, we will load value `10` from database to `self.component_nonce`.
And whenever a new nonce gets created, we will repeat the procedure above. And we will assign another component to the same id.

### PR DESCRIPTION
**Pull request by pmnoxx**
Closes #5133 

After we remove edges we store them inside components, each with unique `component_nonce`. 
When that happens, we store that `component_nonce` on disk, increase it by 1 in memory, and then create new component with `component_nonce`.
However, if the node restarts we will load `component_nonce` value from disk, and when we remove edges, we will create another component with `component_nonce`.

That's a problem, because ids of components are no longer unique, and we will override previously stored component on disk. 

To solve this problem, we will change the logic to;
- use `component_nonce` for current component
- store `component_nonce+1` on disk
- increase `component_nonce` for future

### CODE REVIEW

File: chain/network/src/routing.rs [hunk 1]
```diff
@@ -769,23 +769,23 @@ impl RoutingTable {
         }
         debug!(target: "network", "try_save_edges: We are going to remove {} peers", to_save.len());
 
-        let component_nonce = self.component_nonce;
+        let old_component_nonce = self.component_nonce;
         self.component_nonce += 1;
 
         let mut update = self.store.store_update();
-        let _ = update.set_ser(ColLastComponentNonce, &[], &component_nonce);
+        let _ = update.set_ser(ColLastComponentNonce, &[], &self.component_nonce);
```
Comment (bowenwang1996) [inline chain/network/src/routing.rs:776]: It seems that we are not actually storing last component nonce in the database, since that would be `old_component_nonce`. Should we rename the column?

**Review by bowenwang1996 — COMMENTED** [review]
It would be beneficial to add in your PR description how this change fixes the issue you mentioned.

File: chain/network/src/routing.rs (hunk 1 above)
Comment (pmnoxx) [reply chain/network/src/routing.rs:776]: That's a good observation. Please take a look at https://github.com/near/nearcore/pull/5140/files and https://github.com/near/nearcore/pull/5089
I renamed this variable there to have the proper meaning.

**Comment by pmnoxx** [comment]
> It would be beneficial to add in your PR description how this change fixes the issue you mentioned.

@bowenwang1996 I updated the description of the fchange.

**Comment by pmnoxx** [comment]
@mm-near Can you take a look?

File: chain/network/src/routing.rs [hunk 2]
```diff
@@ -769,23 +769,23 @@ impl RoutingTable {
         }
         debug!(target: "network", "try_save_edges: We are going to remove {} peers", to_save.len());
 
-        let component_nonce = self.component_nonce;
+        let old_component_nonce = self.component_nonce;
```
Comment (mm-near) [inline chain/network/src/routing.rs:772]: I think that the original code didn't have the bug, but was very confusing..

If you look at the line 428:
self.component_nonce = 
.map_or(0, |nonce| nonce + 1);

So basically the approach was:
- the "nonce" that is stored in the database, is the last row/nonce that was used to store something there.
Then - on startup (like 428) - we read it from DB and increment.

which means that self.component_nonce is the next nonce that can be used.

**Review by mm-near — CHANGES_REQUESTED** [review]
TL;DR: from what I see this is not a bug, but the code is a little bit confusing - and we might want to update the comments instead.

File: chain/network/src/routing.rs (hunk 2 above)
Comment (pmnoxx) [reply chain/network/src/routing.rs:772]: Ok, then we can close this PR.
Renaming `component_nonce` to `next_unused_component_nonce` should be enough.
